src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.IO;
+using System.Threading;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.ProjectSystem;
+
+[Shared]
+[Export(typeof(RazorProjectInfoEndpointPublisher))]
+internal class RazorProjectInfoEndpointPublisher
+{
+    private readonly LSPRequestInvoker _requestInvoker;
+    private bool _useCache;
+
+    private readonly Dictionary<ProjectKey, IProjectSnapshot> _mappings;
```

Comment:
Similar to above, I don't think I understand why this class has its own cache, when the ProjectSnapshotManager already does this job.

It seems like this class could be relatively straight forward: It would have a single "StartSending" method, called when the LSP server has been initialized, and when called, that method iterates through all projects in the ProjectSnapshotManager, and sends them, and then hooks up to the snapshot manager changed event. When changes come in, this class sends updates. If we want debouncing or delaying, this would be the logical place to do it (related to above comments about delaying in the server).

It _does_ mean the event handler in this class would need some of the same logic as the `RazorProjectInfoPublisher` currently has, but I think thats okay because it is subtley different. eg, this class knows it won't do anything until the LSP server is started, so it doesn't have to care about waiting for the first document opens, or the LSPEditorAvailable method etc.

Reply:
Used this suggestion is https://github.com/dotnet/razor/pull/10182/commits/4ae01c82888ec4acc7403ed3f42c4106bd7f426a, thanks, I think it worked well!